e2e_tests/standard/test_files.py

```diff
@@ -121,3 +125,168 @@ def test_fileset(self, container: AttributeContainer):
                         content = file.read()
                         assert len(content) == len(expected_content)
                         assert content == expected_content
+
+    @staticmethod
+    def _gen_possible_paths():
+        dir_lvl1_names = [fake.word() for _ in range(3)]
+        dir_lvl2_names = [fake.word() for _ in range(3)]
+        dir_lvl3_names = [fake.word() for _ in range(3)]
+        return [
+            "/".join(prod)
+            for prod in product(dir_lvl1_names, dir_lvl2_names, dir_lvl3_names)
+        ]
+
+    @pytest.mark.parametrize("container", ["project", "run"], indirect=True)
+    def test_fileset_nested_structure(self, container: AttributeContainer):
+        key = self.gen_key()
+        possible_paths = self._gen_possible_paths()
+
+        small_files = [
+            (
+                f"{random.choice(possible_paths)}/{uuid.uuid4()}.{fake.file_extension()}",
+                fake.sentence().encode("utf-8"),
+            )
+            for _ in range(100)
+        ]
+
+        with tmp_context():
+            # create dirs
+            for dir_path in possible_paths:
+                path = Path(dir_path)
+                path.mkdir(parents=True)
```

Comment:
Instead of these two lines probably `os.makedirs` is more common. https://docs.python.org/3/library/os.html#os.makedirs